Extract function to retrieve web apps available to user into utility

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
This is somewhat of a prefactor in order to facilitate using the same list of app docs when collecting mobile UCRs to include in a restore. Existing behavior should not have changed at all with this PR.

I added automated tests for what felt like the most complex part of the utility, which was determining if a user is able to access an app or not. 

Commit by commit 🐟 
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added tests in corehq.apps.cloudcare.tests.test_utils:TestCanUserAccessWebApp to ensure `can_access_web_app` behaves as expected.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Cleanup method to determine web apps accessibility

gherceg committed:
Run isort on corehq/apps/cloudcare/views.py

gherceg committed:
Move fetch helpers to cloudcare/utils.py

gherceg committed:
Move method to get available web apps for user

MartinRiese commented on corehq/apps/cloudcare/utils.py:
```diff
@@ -10,6 +10,28 @@
+            get_application_access_for_domain,
+        )
+        app_access = get_application_access_for_domain(domain)
+        has_access_via_group = app_access.user_can_access_app(user, app)
```
When will this go away? Isn't that already deprecated on staging?

gherceg replied:
Yeah my understanding is it is deprecated but I have no insight into when it will go away. @orangejenny probably has more insight into that.

orangejenny replied:
It's limited to flagged projects, but it still needs to work for those projects indefinitely. Ticket for truly moving these projects off the flag and removing the code: https://dimagi.atlassian.net/browse/USH-4510

MartinRiese reviewed (approved):
Looks pretty straight forward to me.

millerdev commented on corehq/apps/cloudcare/views.py:
```diff
@@ -109,21 +113,16 @@ def fetch_app(self, domain, app_id):
-        apps = filter(None, apps)
-        apps = filter(lambda app: app.get('cloudcare_enabled') or self.preview, apps)
-        # Backwards-compatibility - mobile users haven't historically required this permission
-        if user.is_web_user() or user.can_access_any_web_apps(domain):
-            apps = filter(lambda app: user.can_access_web_app(domain, app.get('copy_of', app.get('_id'))), apps)
-        if toggles.WEB_APPS_PERMISSIONS_VIA_GROUPS.enabled(domain):
-            apps = filter(lambda app: app_access.user_can_access_app(user, app), apps)
```
I think the new function could result in a different outcome than this if both `if` statements evaluate to true. Not sure if that is possible in practice?

Example: an app would have been eliminated from the list of apps if the app was excluded by the first `filter(...)` but would have been included in the second if it had gotten that far (effectively these two conditions are combined with `and`), whereas the new `can_user_access_web_app` function would return true because those conditions are combined with `or`.

gherceg replied:
Yeah this was a great catch. I believe I've addressed this concern and now give precedence to permissions over groups. There are a couple of other threads in this PR that highlight similar issues. See 92ffbbc2c5c2e55fa3d713b6cc2b891c67413c02

millerdev commented on corehq/apps/cloudcare/views.py:
```diff
@@ -105,27 +107,15 @@ class FormplayerMain(View):
 
-    def fetch_app(self, domain, app_id):
-        return _fetch_build(domain, self.request.couch_user.username, app_id)
+    def fetch_app_fn(self):
```
Maybe make this a property?
```suggestion
    @property
    def fetch_app_fn(self):
```

This will obviously require a few other changes as well if accepted.

gherceg replied:
So I initially did this, and ran into an issue in tests related to attempting to pickle something when this was a property, which I didn't feel like digging into too much since sticking with a method sounded easier.

millerdev commented on corehq/apps/cloudcare/views.py:
```diff
@@ -105,27 +107,15 @@ class FormplayerMain(View):
         apps = [_format_app_doc(app) for app in apps]
-        apps = sorted(apps, key=lambda app: app['name'].lower())
-        return apps
+        return sorted(apps, key=lambda app: app['name'].lower())
```
Unrelated to the changes in this PR: is the result of this function always used in a context where it needs to be sorted (like because is displayed to a user)? If not, it would be good to move the sorting to where it's required since sorting is a relatively expensive operation.

gherceg replied:
Yeah the two uses are [here](https://github.com/dimagi/commcare-hq/blob/6a6c72985c14082b77a75ef05998d39447faf597/corehq/apps/cloudcare/views.py#L176-L183), one adds `apps` into a context used to render a page, and the `get_option_apps` usage appears to be called from our javascript code, so both seem necessary. I specifically kept the sorting in this method and not in the new utility function for the reason you highlighted.

gherceg committed:
Grandfather commcare users into web apps access

orangejenny commented on corehq/apps/cloudcare/tests/test_utils.py:
```diff
@@ -107,3 +119,103 @@ def setUpClass(cls):
+        self.assertTrue(has_access)
+
+    @flag_enabled("WEB_APPS_PERMISSIONS_VIA_GROUPS")
+    def test_commcare_user_has_access_if_in_group(self):
```
Could you add a negative test case?

orangejenny replied:
Also a test case for a web user on a domain that uses this flag?

orangejenny replied:
It's hypothetically possible for permissions to grant access to *different* apps than groups, which would also be good to test because it's not immediately obvious what the behavior *should* be. From the description of https://github.com/dimagi/commcare-hq/pull/34421:

> In addition to [the new app-specific changes], any legacy ApplicationAccess restrictions will continue to be applied, for backwards compatibility.

gherceg replied:
Between 92ffbbc2c5c2e55fa3d713b6cc2b891c67413c02 and a995265e9b2288d0e36e868bd1d75de37a0b00e7 I _think_ I've touched on all of these needed test cases.

orangejenny commented on corehq/apps/cloudcare/utils.py:
```diff
@@ -10,6 +10,28 @@
+        app_access = get_application_access_for_domain(domain)
+        has_access_via_group = app_access.user_can_access_app(user, app)
+
+    return has_access_via_permission or has_access_via_group
```
This is clear, but I think it's inaccurate. Permissions take precedence over the groups-based functionality. One way to address this (which would address my comment above about web users and the groups-based permissions) would be to switch `or` to `and`.

gherceg replied:
Resolved in 92ffbbc2c5c2e55fa3d713b6cc2b891c67413c02. I also had to default `has_access_via_groups` to True in addition to the change from `or` to `and`.

orangejenny reviewed (commented):
Left some comments. I recognize it's unjust that tests here are way overdue, you added a bunch, and in gratitude I...asked you to write a bunch more.

Are you confident this won't change behavior around Login As? There's a table at the bottom of https://github.com/dimagi/commcare-hq/pull/34421 that summarizes the current behavior.

gherceg commented on corehq/apps/cloudcare/utils.py:
```diff
@@ -10,6 +10,28 @@
 from corehq.util.quickcache import quickcache
 
 
+def can_user_access_web_app(app, user, domain):
```
I had rearranged the parameters locally but decided to leave it for the sake of consistency, but glad to know it wasn't just me 😄. I can make this change.

gherceg replied:
Changed in 14f3c6afc04190945b31e0069ec44fbf4fb08971

gherceg commented on corehq/apps/cloudcare/utils.py:
```diff
@@ -6,10 +6,70 @@
+        has_access_via_permission = user.can_access_web_app(domain, app.get('copy_of', app.get('_id')))
+
+    has_access_via_group = False
+    if toggles.WEB_APPS_PERMISSIONS_VIA_GROUPS.enabled(domain):
```
> That'll make has_access_via_group always true for web users in domains with this flag on

I assume this isn't a problem in isolation (and is actually correct), but that because of the `or` logic below (`has_access_via_permission or has_access_via_group`), this is incorrect.

Previously the apps that a web user didn't have access to due to permissions would have been filtered out before getting to this code, whereas now the function attempts to check both at the same time. So the only time the group permission would have applied previously was when an app was **not** filtered out in the permissions filter (i.e., has_access_via_permission is true) and then the user did not have access via groups, so then the app was filtered out in that step.

All that to say, I believe this is the same precedence related change you highlighted above (`has_access_via_permission and has_access_via_group`) right?

orangejenny replied:
yes, that's right

gherceg replied:
Should be resolved in 92ffbbc2c5c2e55fa3d713b6cc2b891c67413c02

MartinRiese commented:
Will this touch on the "login as" workflow? I just noticed that is possible to login as a web user and gain access to an app that you are not able to see when logging in directly. I am pretty certain that I had fixed that with the first PR that allowed to login as a web user.

gherceg committed:
Permissions take precedence over groups

I also added more test cases generally here

gherceg commented:
> Will this touch on the "login as" workflow? I just noticed that is possible to login as a web user and gain access to an app that you are not able to see when logging in directly. I am pretty certain that I had fixed that with the first PR that allowed to login as a web user.

If you are referring to behavior you just observed on staging, the logic wasn't quite correct in my change, so it was possible that a web user could see apps they didn't have access to since the `has_access_via_groups` permission always evaluated to True, and I `or`ed the two conditions together. I just pushed a change that should resolve this, and am deploying that to staging now.

gherceg committed:
Cleanup tests

- create helpers for assigning roles and groups
- ensure names are consistent
- add any obvious missing test cases

gherceg committed:
Remove domain parameter since it can be found on app doc

gherceg commented:
Noting that Jenny and I offlined about the comment:
> Are you confident this won't change behavior around Login As? There's a table at the bottom of https://github.com/dimagi/commcare-hq/pull/34421 that summarizes the current behavior.
And decided that this change shouldn't have a direct impact on that logic since it is strictly attempting to replicate the existing behavior here, which in this context is unaware of the Login As user. It simply says "given a user, here are the available web apps".

orangejenny commented on corehq/apps/cloudcare/utils.py:
```diff
@@ -6,10 +6,70 @@
+    app_ids = get_app_ids_in_domain(domain)
+    for app_id in app_ids:
+        app = fetch_app_fn(domain, user.username, app_id)
+        if app and is_web_app(app) and can_user_access_web_app(user, app):
```
Could you also call this new function in `FormplayerPreviewSingleApp.get`? It currently has some of the logic that you've so nicely encapsulated.

gherceg replied:
Done in 4a68aef15fcd1818f42fe3a0fb1d27ac567bf644

millerdev commented on corehq/apps/cloudcare/utils.py:
```diff
@@ -17,12 +17,12 @@
-    :param domain: domain name
+    :param app: app doc (not wrapped Application)
     """
+    domain = app.get("domain")
```
I think this should never be absent, and we probably want an error if it is?
```suggestion
    domain = app["domain"]
```

gherceg replied:
Yeah that sounds right to me. Changed in 4a68aef15fcd1818f42fe3a0fb1d27ac567bf644

gherceg committed:
Use can_user_access_web_app in FormplayerPreviewSingleApp

gherceg committed:
Address feedback in can_user_access_web_app

- raise KeyError if domain key not found on app
- move app_id calculation to its own line

gherceg committed:
Fix missing self parameter

gherceg commented on corehq/apps/cloudcare/utils.py:
```diff
@@ -6,10 +6,70 @@
+    # Backwards-compatibility - mobile users haven't historically required this permission
+    has_access_via_permission = False
+    if user.is_web_user() or user.can_access_any_web_apps(domain):
+        has_access_via_permission = user.can_access_web_app(domain, app.get('copy_of', app.get('_id')))
```
See 0ddf5addceae5d9cd68b87172db60d33c405e745

gherceg replied:
I resolved this prematurely. I don't have a good answer to the possibility of `app_id` being None. I'd imagine it makes sense to change `app.get('_id')` to `app['_id']` so that it fails hard.

gherceg committed:
Use app_json in _format_app_doc
